core/src/main/scala/org/apache/spark/NewAccumulator.scala

```diff
@@ -0,0 +1,356 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark
+
+import java.{lang => jl}
+import java.io.{ObjectInputStream, ObjectOutputStream}
+import java.util.concurrent.atomic.AtomicLong
+import javax.annotation.concurrent.GuardedBy
+
+import org.apache.spark.scheduler.AccumulableInfo
+import org.apache.spark.util.Utils
+
+
+private[spark] case class AccumulatorMetadata(
+    id: Long,
+    name: Option[String],
+    countFailedValues: Boolean) extends Serializable
+
+trait UpdatedValue extends Serializable
```

Comment:
cc @rxin , I didn't send the accumulator back for a serialization problem.

Basically when we send accumulator from driver to executors, we don't want to send its current value(think about list accumulator, we definitely don't wanna send the current list to executors.).
But when we send accumulator from executors to driver, we do need to send the current value.

One possible solution is to have 2 local variables for each accumulator, one for driver, one for executors. But it's a lot of trouble when accumulators have complex intermedia type, e.g. average accumulator. So I end up with this apporach.